File: session-1/nodejs/consumer.js
/**
 * Consumer RabbitMQ — Écoute la queue "capteurs" et affiche les messages reçus.
 */

const amqp = require("amqplib");

async function main() {
  // Connexion au serveur RabbitMQ (localhost par défaut)
  const connection = await amqp.connect("amqp://localhost");
  const channel = await connection.createChannel();

  // Nom de la queue
  const queue = "capteurs";

  // Déclaration de la queue "capteurs"
  // On la déclare aussi côté consumer pour être sûr qu'elle existe
  await channel.assertQueue(queue);

  console.log("[*] En attente de messages. Ctrl+C pour quitter");

  // Abonnement à la queue "capteurs"
  // noAck: true — les messages sont automatiquement acquittés (simplification session 1)
  channel.consume(
    queue,
    (msg) => {
      if (msg !== null) {
        // Décodage du message JSON
        const message = JSON.parse(msg.content.toString());

        // Extraction des données
        const room = message.room || "inconnu";
        const sensor = message.sensor || "inconnu";
        const value = message.value || 0;

        // Affichage formaté selon le type de capteur
        if (sensor === "temperature") {
          console.log(`[${room}] Temperature: ${value}°C`);
        } else if (sensor === "humidity") {
          console.log(`[${room}] Humidite: ${value}%`);
        } else {
          console.log(`[${room}] ${sensor}: ${value}`);
        }
      }
    },
    { noAck: true }
  );

  // Gestion de l'arrêt propre avec Ctrl+C
  process.on("SIGINT", async () => {
    console.log("\n[*] Arrêt du consumer");
    await channel.close();
    await connection.close();
    process.exit(0);
  });
}

// Lancement du consumer
main().catch(console.error);

File: session-1/python/consumer.py
#!/usr/bin/env python3
"""
Consumer RabbitMQ — Écoute la queue "capteurs" et affiche les messages reçus.
"""

import pika
import json

# Connexion au serveur RabbitMQ (localhost par défaut)
connection = pika.BlockingConnection(pika.ConnectionParameters('localhost'))
channel = connection.channel()

# Déclaration de la queue "capteurs"
# On la déclare aussi côté consumer pour être sûr qu'elle existe
# (peu importe qui démarre en premier, producer ou consumer)
channel.queue_declare(queue='capteurs')


def callback(ch, method, properties, body):
    """Fonction appelée à chaque message reçu."""
    # Décodage du message JSON
    message = json.loads(body)

    # Extraction des données
    room = message.get("room", "inconnu")
    sensor = message.get("sensor", "inconnu")
    value = message.get("value", 0)

    # Affichage formaté selon le type de capteur
    if sensor == "temperature":
        print(f"[{room}] Temperature: {value}°C")
    elif sensor == "humidity":
        print(f"[{room}] Humidite: {value}%")
    else:
        print(f"[{room}] {sensor}: {value}")


# Abonnement à la queue "capteurs"
# auto_ack=True : les messages sont automatiquement acquittés (simplification session 1)
channel.basic_consume(
    queue='capteurs',
    on_message_callback=callback,
    auto_ack=True
)

print("[*] En attente de messages. Ctrl+C pour quitter")

try:
    # Lancement de l'écoute (boucle infinie)
    channel.start_consuming()
except KeyboardInterrupt:
    print("\n[*] Arrêt du consumer")
    channel.stop_consuming()

connection.close()
